app/Parsers/RequestHeaderParser/RequestHeaderParser.php

```diff
@@ -0,0 +1,171 @@
+     *
+     * @var array
+     */
+    private static $allowedMethods = [
+        'GET',
+        'HEAD',
+        'POST',
+        'PUT',
+        'DELETE',
+        'CONNECT',
+        'OPTIONS',
+        'TRACE',
+        'PATCH',
+    ];
+
+    /**
+     * Method that should be used to execute the HTTP request.
+     *
+     * @var string
+     */
+    private $method;
+
+    /**
+     * Url that should be used to execute the HTTP request.
+     *
+     * @var string
+     */
+    private $url;
+
+    /**
+     * The response headers from the HTTP request.
+     *
+     * @var array
+     */
+    private $headers;
+
+    /**
+     * @param $method
+     * @throws InvalidEndpointMethodException
+     */
+    public function setMethod($method)
+    {
+        if (! in_array($method, static::$allowedMethods)) {
+            throw new InvalidEndpointMethodException();
+        }
+
+        $this->method = $method;
+    }
+
+    /**
+     * @param $url
+     * @throws InvalidEndpointUrlException
+     */
+    public function setUrl($url)
+    {
+        if (! filter_var($url, FILTER_VALIDATE_URL)) {
+            throw new InvalidEndpointUrlException();
+        }
+
+        $this->url = $url;
+    }
+
+    /**
+     * Execute request and parse responses with headers.
+     *
+     * @return Response[]
+     */
+    public function getResponses(): array
+    {
+        $this->execute();
+
+        /* @var Response[] $responses */
+        $responses = [];
+        foreach ($this->headers as $headerName => $headerValue) {
+            if (! is_int($headerName)) {
+                continue;
+            }
+
+            $response = new Response;
+            $response->setStatusCode($this->getStatusCode($headerValue));
+            $response->setReasonPhrase($headerValue);
+
+            array_push($responses, $response);
+        }
+
+        $currentResponseIndex = 0;
+        foreach ($this->headers as $headerName => $headerValue) {
+            if (is_int($headerName)) {
+                $currentResponseIndex = $headerName;
+
+                continue;
+            }
+
+            if (is_array($headerValue)) {
+                foreach ($headerValue as $responseIndex => $value) {
+                    $header = new Header;
+                    $header->setName($headerName);
+                    $header->setValue($value);
+
+                    $responses[$responseIndex]->addHeader($header);
+                }
+            } else {
+                $header = new Header;
+                $header->setName($headerName);
+                $header->setValue($headerValue);
+
+                $responses[$currentResponseIndex]->addHeader($header);
+            }
+        }
+
+        return $responses;
+    }
+
+    /**
+     * Get status code by HTTP reason phrase.
+     *
+     * @param $reasonPhrase
+     * @return int
+     */
+    private function getStatusCode($reasonPhrase): int
+    {
+        preg_match('/[0-9]{3}/', $reasonPhrase, $matches);
```

Comment:
This regular expression should be stricter. Please only accept `1-5` as first character/number.